src/ucp/rndv/rndv_ppln.c

```diff
@@ -84,43 +85,51 @@ ucp_proto_rndv_ppln_probe(const ucp_proto_init_params_t *init_params)
         return;
     }
 
-    /* Find the performance range of sending one fragment */
-    if (!ucp_proto_select_get_valid_range(select_elem->thresholds,
-                                          &frag_min_length, &frag_max_length)) {
-        return;
-    }
+    /* Add each proto as a separate variant */
+    ucs_array_for_each(proto, &select_elem->proto_init.protocols) {
+        /* Skip empty variants and reconfig proto */
+        if ((proto->caps.num_ranges == 0) || (proto->proto_id == 0)) {
+            continue;
+        }
 
-    /* Initialize private data */
-    rpriv.frag_proto = *select_elem;
-    rpriv.frag_size  = frag_max_length;
-
-    frag_range  = ucp_proto_perf_range_search(select_elem, frag_max_length);
-    thresh_elem = ucp_proto_select_thresholds_search(select_elem,
-                                                     frag_max_length);
-
-    ucs_trace("rndv_ppln frag %s" UCP_PROTO_PERF_FUNC_TYPES_FMT,
-              ucs_memunits_to_str(rpriv.frag_size, frag_size_str,
-                                  sizeof(frag_size_str)),
-              UCP_PROTO_PERF_FUNC_TYPES_ARG(frag_range->perf));
-
-    /* Add the single range of the pipeline protocol to ppln_caps */
-    ppln_caps.min_length = frag_max_length + 1;
-    ppln_caps.num_ranges = 0;
-    ucp_proto_common_add_ppln_range(&ppln_caps, frag_range, SIZE_MAX);
-
-    /* Add ATS overhead */
-    ppln_overhead = ucs_linear_func_make(frag_overhead,
-                                         frag_overhead / frag_max_length);
-    status = ucp_proto_rndv_ack_init(init_params, UCP_PROTO_RNDV_ATS_NAME,
-                                     &ppln_caps, ppln_overhead, &rpriv.ack,
-                                     &caps);
-    if (status == UCS_OK) {
-        ucp_proto_select_add_proto(init_params,
-                                   thresh_elem->proto_config.cfg_thresh, 0,
-                                   &caps, &rpriv, sizeof(rpriv));
-    }
+        frag_range = proto->caps.ranges + (proto->caps.num_ranges - 1);
```

Comment:
`ucp_proto_perf_range_search` is supposed to use in select_elem and the current logic just gets the last range from caps. We cannot search through `select_elem` since it contains envelope and we need to handle only one certain variant here.

Reply:
ok